### DISCUSSION TITLE
Parsing parameter file failed at force-level2

### DISCUSSION THREAD

**Discussion by cxxuuuuuuu** [davidfrantz/force#102] [Q&A] [answered]
Hi all,

I got an error when I run ''force-level2 /data/force/param/l2ps.prm'':

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
ETA: 1907s Left: 587 AVG: 3.25s  local:1/279/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1904s Left: 586 AVG: 3.25s  local:1/280/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1900s Left: 585 AVG: 3.25s  local:1/281/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1896s Left: 584 AVG: 3.25s  local:1/282/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1893s Left: 583 AVG: 3.25s  local:1/283/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1890s Left: 582 AVG: 3.25s  local:1/284/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1887s Left: 581 AVG: 3.25s  local:1/285/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1883s Left: 580 AVG: 3.25s  local:1/286/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1880s Left: 579 AVG: 3.25s  local:1/287/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1877s Left: 578 AVG: 3.25s  local:1/288/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1874s Left: 577 AVG: 3.25s  local:1/289/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1871s Left: 576 AVG: 3.26s  local:1/290/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1868s Left: 575 AVG: 3.25s  local:1/291/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 1865s Left: 574 AVG: 3.25s  local:1/292/100%/3.3s No valid parameter file!
Parsing parameter file failed.

...

I have 866 Sentinel 2 level-1C files. It worked for some files at the beginning. Then the other files could not be processed as showed in the error.

Here is my parameter file:

++PARAM_LEVEL2_START++

# INPUT/OUTPUT DIRECTORIES
# ------------------------------------------------------------------------
FILE_QUEUE = /home/eouser/level-1-datapool/l1.pool
DIR_LEVEL2 = /home/eouser/src/force/level2
DIR_LOG = /home/eouser/src/force/log
DIR_TEMP = /home/eouser/src/force/temp

# DIGITAL ELEVATION MODEL
# ------------------------------------------------------------------------
FILE_DEM = NULL
DEM_NODATA = -32767

# DATA CUBES
# ------------------------------------------------------------------------
DO_REPROJ = TRUE
DO_TILE = TRUE
FILE_TILE = /home/eouser/src/force/misc/bw.txt
TILE_SIZE = 30000
BLOCK_SIZE = 3000
RESOLUTION_LANDSAT = 30
RESOLUTION_SENTINEL2 = 10

ORIGIN_LON = 7.92
ORIGIN_LAT = 51.51
PROJECTION = PROJCS["ETRS89 / LAEA Europe",GEOGCS["ETRS89",DATUM["European_Terrestrial_Reference_System_1989",SPHEROID["GRS 1980",6378137,298.257222101,
AUTHORITY["EPSG","7019"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY["EPSG","6258"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],
UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4258"]],PROJECTION["Lambert_Azimuthal_Equal_Area"],
PARAMETER["latitude_of_center",52],PARAMETER["longitude_of_center",10],PARAMETER["false_easting",4321000],PARAMETER["false_northing",3210000],
UNIT["metre",1,AUTHORITY["EPSG","9001"]],AUTHORITY["EPSG","3035"]]

RESAMPLING = CC

# RADIOMETRIC CORRECTION OPTIONS
# ------------------------------------------------------------------------
DO_ATMO = TRUE
DO_TOPO = TRUE
DO_BRDF = TRUE
ADJACENCY_EFFECT = TRUE
MULTI_SCATTERING = TRUE

# WATER VAPOR CORRECTION OPTIONS
# ------------------------------------------------------------------------
DIR_WVPLUT = NULL
WATER_VAPOR = NULL

# AEROSOL OPTICAL DEPTH OPTIONS
# ------------------------------------------------------------------------
DO_AOD  = TRUE
DIR_AOD  = NULL

# CLOUD DETECTION OPTIONS
# ------------------------------------------------------------------------
ERASE_CLOUDS = FALSE
MAX_CLOUD_COVER_FRAME = 90
MAX_CLOUD_COVER_TILE  = 100
CLOUD_BUFFER  = 300
SHADOW_BUFFER = 90
SNOW_BUFFER   = 30
CLOUD_THRESHOLD  = 0.225
SHADOW_THRESHOLD = 0.02

# RESOLUTION MERGING
# ------------------------------------------------------------------------
RES_MERGE = IMPROPHE

# CO-REGISTRATION OPTIONS
# ------------------------------------------------------------------------
DIR_COREG_BASE = NULL
OREG_BASE_NODATA = -9999

# MISCELLANEOUS OPTIONS
# ------------------------------------------------------------------------
MPULSE_NOISE = TRUE
BUFFER_NODATA = FALSE

# TIER LEVEL
# ------------------------------------------------------------------------
TIER = 1

# PARALLEL PROCESSING
# ------------------------------------------------------------------------
NPROC = 32
NTHREAD = 2
PARALLEL_READS = FALSE
DELAY = 3

# OUTPUT OPTIONS
# ------------------------------------------------------------------------
OUTPUT_FORMAT = GTiff
OUTPUT_DST = FALSE
OUTPUT_AOD = FALSE
OUTPUT_WVP = FALSE
OUTPUT_VZN = FALSE
OUTPUT_HOT = FALSE
OUTPUT_OVV = TRUE

++PARAM_LEVEL2_END++

Does anybody know a solution? Thanks a lot.

Best regards,
Chunyan

**Comment by JonasardoLundUniv** [comment]
Hi

I don't think **DO_TOPO = TRUE** work with  **FILE_DEM = NULL**
DEM needed to do topo-correction.

Regards
/Jonas

**Reply by cxxuuuuuuu** [reply]
Hi @JonasardoLundUniv ,

thank you! I have changed DO_TOPO = TRUE to DO_TOPO = FALSE.

Regards,
Chunyan

**Comment by davidfrantz** [comment]
Hi,

yes, @JonasardoLundUniv is right. However, the error message here is because of some other error that appears before the DEM error message would have been emitted.

Probably very much related to this thread here:
https://github.com/davidfrantz/force/discussions/84

This is a strange error that keeps popping up, and I am not able to reproduce so far.

What FORCE version are you using?
What data are you processing? Landsat / S2?
What System are you running on? Do you have enough RAM etc.?
Are you using Docker?

Can you please attach the parameter file (when you post the content only, I cannot see if something is wrong with the file itself).

Cheers,
David

**Reply by cxxuuuuuuu** [reply]
Hi David,

thank you for the answer. I am using FORCE 3.6.5, and only processing Sentinel 2 Level 1C now. 
I run on Windows 10 with a virtual machine to Ubuntu 18.04.5 LTS (GNU/Linux 4.15.0-124-generic x86_64) where the RAM is 32 GB. 
And I am not using Docker. 

Could you please tell me how to attach a file here? Because I couldn't find the option to upload my parameter file. Thanks!

Cheers,
Chunyan

**Comment by davidfrantz** [comment, marked as answer]
Hi Chunyan,
I believe you can simply drag'n'drop a file when writing a message.
Cheers,
David

**Reply by cxxuuuuuuu** [reply]
Thanks! Here is my parameter file:

[12ps.prm.txt](https://github.com/davidfrantz/force/files/6569102/12ps.prm.txt)

**Reply by davidfrantz** [reply]
Thanks,
on first sight, I do not see any obvious formatting issue.
However, I believe that you may simply exhausting RAM. Have you monitored RAM usage?
You are using a 32GB VM, and are using ``NPROC = 32`` processes.
Probably, all these error messages are related to insufficient RAM.
Can you please have a read through this thread, and see whether it solves your issue?
https://github.com/davidfrantz/force/discussions/94
Cheers,
David

**Reply by cxxuuuuuuu** [reply]
Hi David,

thanks for your advice. I changed my virtual machine to have a RAM of 64 GB. At the same time, I narrowed down the range of data which only contains 26 images for one month (20180801,20180831). 

However, no matter how many times I try, **the same error is still there**:

eouser@force:~$ force-level2 /home/eouser/src/force/param/12ps.prm
26 images enqueued. Start processing with 32 CPUs
No valid parameter file!
Parsing parameter file failed.
Computers / CPU cores / Max jobs to run
1:local / 8 / 26

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
ETA: 0s Left: 26 AVG: 0.00s  local:26/0/100%/0.0s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15540 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 84s Left: 25 AVG: 13.00s  local:25/1/100%/408.0s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15787 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 267s Left: 24 AVG: 16.50s  local:24/2/100%/214.0s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15478 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 288s Left: 23 AVG: 14.00s  local:23/3/100%/145.7s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 16063 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 256s Left: 22 AVG: 12.25s  local:22/4/100%/111.0s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15631 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 221s Left: 21 AVG: 10.80s  local:21/5/100%/89.8s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15694 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 191s Left: 20 AVG: 9.83s  local:20/6/100%/75.7s sh: error while loading shared libraries: libc.so.6: cannot create shared object descriptor: Cannot allocate memory
ETA: 197s Left: 20 AVG: 10.17s  local:20/6/100%/76.0s No valid parameter file!
Parsing parameter file failed.
ETA: 186s Left: 19 AVG: 10.14s  local:19/7/100%/66.6s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15926 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 166s Left: 18 AVG: 9.38s  local:18/8/100%/58.8s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 15849 Killed                  $EXE $TODO $PRM >> $LOGFILE


**and I got new errors:**

ETA: 152s Left: 17 AVG: 9.11s  local:17/9/100%/53.0s No valid parameter file!
Parsing parameter file failed.
CPLCreateJoinableThread() failed: Resource temporarily unavailable.
CPLCreateJoinableThread() failed: Resource temporarily unavailable.
CPLCreateJoinableThread() failed: Resource temporarily unavailable.
CPLCreateJoinableThread() failed: Resource temporarily unavailable.
ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1008738336) in GDALRasterBand::TryGetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1008738336) in GDALRasterBand::GetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1701867372) in GDALRasterBand::TryGetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1701867372) in GDALRasterBand::GetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1684632135) in GDALRasterBand::TryGetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (1684632135) in GDALRasterBand::GetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (538976288) in GDALRasterBand::TryGetLockedBlockRef()

ERROR 5: T32UMC_20180828T103019_B09.jp2, band 1: Illegal nBlockXOff value (538976288) in GDALRasterBand::GetLockedBlockRef()

ETA: 150s Left: 16 AVG: 9.50s  local:16/10/100%/49.0s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 17303 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 199s Left: 15 AVG: 13.45s  local:15/11/100%/49.4s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 16362 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 184s Left: 14 AVG: 13.25s  local:14/12/100%/46.2s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 16979 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 160s Left: 13 AVG: 12.31s  local:13/13/100%/42.7s No valid parameter file!
Parsing parameter file failed.
ETA: 158s Left: 12 AVG: 13.29s  local:12/14/100%/41.5s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 16157 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 192s Left: 11 AVG: 17.53s  local:11/15/100%/43.9s No valid parameter file!
Parsing parameter file failed.
/usr/local/bin/force-l2ps_: line 161: 16580 Killed                  $EXE $TODO $PRM >> $LOGFILE
ETA: 241s Left: 10 AVG: 24.12s  local:10/16/100%/48.8s No valid parameter file!
Parsing parameter file failed.
ETA: 249s Left: 9 AVG: 27.71s  local:9/17/100%/50.9s No valid parameter file!
Parsing parameter file failed.
ETA: 223s Left: 8 AVG: 28.00s  local:8/18/100%/49.9s No valid parameter file!
Parsing parameter file failed.
ETA: 218s Left: 7 AVG: 31.26s  local:7/19/100%/52.1s No valid parameter file!
Parsing parameter file failed.
ETA: 332s Left: 6 AVG: 55.45s  local:6/20/100%/75.2s No valid parameter file!
Parsing parameter file failed.
ETA: 268s Left: 5 AVG: 53.67s  local:5/21/100%/72.5s No valid parameter file!
Parsing parameter file failed.
ETA: 217s Left: 4 AVG: 54.36s  local:4/22/100%/72.3s No valid parameter file!
Parsing parameter file failed.
ETA: 156s Left: 3 AVG: 52.00s  local:3/23/100%/69.2s No valid parameter file!
Parsing parameter file failed.
ETA: 99s Left: 2 AVG: 49.92s  local:2/24/100%/66.4s No valid parameter file!
Parsing parameter file failed.
ETA: 343s Left: 1 AVG: 343.80s  local:1/25/100%/359.6s No valid parameter file!
Parsing parameter file failed.


It's so weird. I guess something goes wrong in the parameter file. For example it shows: 

/usr/local/bin/force-l2ps_: line 161: 17303 killed
/usr/local/bin/force-l2ps_: line 161: 16362 killed
/usr/local/bin/force-l2ps_: line 161: 16580 killed
...


But I don't know how to correct them... 

(Parameter file attached) 

[12ps.prm.txt](https://github.com/davidfrantz/force/files/6618419/12ps.prm.txt)


Regards,
Chunyan

**Reply by JonasardoLundUniv** [reply]
Try with 1 CPU instead of 32.

**Reply by cxxuuuuuuu** [reply]
> Try with 1 CPU instead of 32.

Hi @JonasardoLundUniv , 

"Start processing with 32 CPUs" is default here. Could you please tell me how to try 1 CPU? How can I change 32 CPUs to 1 CPU?

Thanks!

Best regards,
Chunyan

**Reply by davidfrantz** [reply]
there is the parameter ``NPROC`` in the parameter file...
Best,
David

**Reply by cxxuuuuuuu** [reply]
> there is the parameter `NPROC` in the parameter file...
> Best,
> David

Hi David,

I changed NPROC to 3 and the ''new errors'' that I mentioned above have been solved. During these two weeks I work very intensively at modifying the parameter file, where I use SRTM as DEM to rectify the Sentinel-2 Level 1 images. But it really doesn't work no matter how many times that I try. 

I have communicated a lot with @skoenig-eo Simon König who also got the similar error at **parsing parameter file**. But we really couldn't find a solution.  #84

The error is still the same:

eouser@force:~$ force-level2 /home/eouser/src/force/param/12ps.prm
13 images enqueued. Start processing with 3 CPUs
No valid parameter file!
Parsing parameter file failed.

Computers / CPU cores / Max jobs to run
1:local / 8 / 3

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
ETA: 0s Left: 13 AVG: 0.00s  local:1/0/100%/0.0s No valid parameter file!
Parsing parameter file failed.
ETA: 36s Left: 12 AVG: 3.00s  local:1/1/100%/4.0s No valid parameter file!
Parsing parameter file failed.
ETA: 30s Left: 11 AVG: 3.00s  local:1/2/100%/3.5s No valid parameter file!
Parsing parameter file failed.
ETA: 27s Left: 10 AVG: 3.00s  local:1/3/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 24s Left: 9 AVG: 3.00s  local:1/4/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 21s Left: 8 AVG: 3.00s  local:1/5/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 20s Left: 7 AVG: 3.17s  local:1/6/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 17s Left: 6 AVG: 3.14s  local:1/7/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 15s Left: 5 AVG: 3.12s  local:1/8/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 12s Left: 4 AVG: 3.11s  local:1/9/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 9s Left: 3 AVG: 3.10s  local:1/10/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 6s Left: 2 AVG: 3.18s  local:1/11/100%/3.3s No valid parameter file!
Parsing parameter file failed.
ETA: 3s Left: 1 AVG: 3.17s  local:1/12/100%/3.2s No valid parameter file!
Parsing parameter file failed.
ETA: 0s Left: 0 AVG: 2.92s  local:0/13/100%/3.0s

[12ps.prm.txt](https://github.com/davidfrantz/force/files/6703009/12ps.prm.txt)

Would you like to check and run the 12ps.prm again and see whether there are errors of parameters? Since not just I got this error, it is very meaningful for helping other FORCE users if there is a solution. Thanks again.


Best regards,
Chunyan

**Reply by davidfrantz** [reply]
Hi @cxxuuuuuuu ,

there are several issues intermingled in this thread.

1) I am glad that you fixed your initial issues with the resource allocation. which caused many of the images to fail. (both because of  overutilization of threads and RAM limitations). This is not the same issue as https://github.com/davidfrantz/force/discussions/84
I would be happy if you could mark my answer above as solution as this was the main issue here.

2) Your current issue is because your parameter file is indeed invalid, probably because you did a lot of testing and somwthing went wrong on the way.

```
tail ~/temp/cxuuuuu/12ps.prm.txt
# Type: Logical. Valid values: {TRUE,FALSE}
OUTPUT_VZN = FALSE
# Output the  Haze Optimzed Transformation output? This product can be
# used in force-level3; no other higher-level FORCE module is using this.
# Type: Logical. Valid values: {TRUE,FALSE}
OUTPUT_HOT = FALSE
# Output overview thumbnails? These are jpegs at reduced spatial resolution,
# which feature an RGB overview + quality information overlayed (pink: cloud,
# red: cirrus, cyan: cloud shadow, yellow: snow, orange: saturated, green:
# subzero reflectance). No higher-leve /codede/auxdata/CopDEM/COP-DEM_GLO-90-DGED/DEM1_SAR_DGE_90_20110408T171540_20130501T171033_ADS_000000_6951.DEM/Copernicus_DSM_30_N51_00_E008_00/DEM/Copernicus_DSM_30_N51_00_E008_00_DEM.ti+
```

should look like this:

```
tail ~/temp/cxuuuuu/LEVEL2-skeleton.prm
# Type: Logical. Valid values: {TRUE,FALSE}
OUTPUT_HOT = FALSE
# Output overview thumbnails? These are jpegs at reduced spatial resolution,
# which feature an RGB overview + quality information overlayed (pink: cloud,
# red: cirrus, cyan: cloud shadow, yellow: snow, orange: saturated, green:
# subzero reflectance). No higher-level FORCE module is using this.
# Type: Logical. Valid values: {TRUE,FALSE}
OUTPUT_OVV = TRUE

++PARAM_LEVEL2_END++
```

Note the last line. After fixing this, it worked as expected on my end.

3) Regarding https://github.com/davidfrantz/force/discussions/84, I accidentally was able to reproduce this strange behaviour when testing your parameter file. I will answer in https://github.com/davidfrantz/force/discussions/84.

Cheers,
David

**Reply by cxxuuuuuuu** [reply]
Hi David,

thanks a lot!! It works now on my end!! Although I got the same error ''Parsing parameter file failed'', but I found that the BOA results were somehow still generated under the /level2 folder, what a incredible thing!!

Thank you for your kind explanation again. 

Best regards from Berlin,
Chunyan

**Comment by davidfrantz** [comment]
Hi @cxxuuuuuuu ,

I am happy to hear that it works on your end now. 

Can you please mark this thread as ``answered``?

For the strange warning, please see my response here: https://github.com/davidfrantz/force/discussions/84

Cheers,
David

**Reply by cxxuuuuuuu** [reply]
Hey David,

yes you can mark it as answered.

Best regards,
Chunyan
